cmd/katapult/mock_api_client.go

```diff
@@ -0,0 +1,215 @@
+package main
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"errors"
+	"io/ioutil"
+	"net/http"
+	"strings"
+
+	"github.com/krystal/go-katapult"
+	"github.com/krystal/go-katapult/core"
+)
+
+var dcs = []*core.DataCenter{
+	{
+		ID:        "POG1",
+		Name:      "hello",
+		Permalink: "Hello World!",
+		Country: &core.Country{
+			ID:   "POG",
+			Name: "Pogland",
+		},
+	},
+	{
+		ID:        "GB1",
+		Name:      "hello",
+		Permalink: "Hello World!",
+		Country: &core.Country{
+			ID:   "UK",
+			Name: "United Kingdom",
+		},
+	},
+}
+
+var idNetworks = []*core.Network{
+	{
+		ID:        "pognet",
+		Name:      "Pognet 1",
+		Permalink: "pog-1",
+		DataCenter: &core.DataCenter{
+			ID:        "POG1",
+			Name:      "Pogland 1",
+			Permalink: "pog1",
+			Country: &core.Country{
+				ID:   "pog",
+				Name: "Pogland",
+			},
+		},
+	},
+	{
+		ID:        "pognet2",
+		Name:      "Pognet 2",
+		Permalink: "pog-2",
+		DataCenter: &core.DataCenter{
+			ID:        "POG1",
+			Name:      "Pogland 1",
+			Permalink: "pog1",
+			Country: &core.Country{
+				ID:   "pog",
+				Name: "Pogland",
+			},
+		},
+	},
+}
+
+var subdomainNetworks = []*core.Network{
+	{
+		ID:        "pognet3",
+		Name:      "Pognet 3",
+		Permalink: "pog-3",
+		DataCenter: &core.DataCenter{
+			ID:        "POG1",
+			Name:      "Pogland 1",
+			Permalink: "pog1",
+			Country: &core.Country{
+				ID:   "pog",
+				Name: "Pogland",
+			},
+		},
+	},
+	{
+		ID:        "pognet4",
+		Name:      "Pognet 4",
+		Permalink: "pog-4",
+		DataCenter: &core.DataCenter{
+			ID:        "POG1",
+			Name:      "Pogland 1",
+			Permalink: "pog1",
+			Country: &core.Country{
+				ID:   "pog",
+				Name: "Pogland",
+			},
+		},
+	},
+}
+
+var organizations = []*core.Organization{
+	{
+		ID:        "loge",
+		Name:      "Loge Enthusiasts",
+		SubDomain: "loge",
+	},
+	{
+		ID:        "testing",
+		Name:      "testing, testing, 123",
+		SubDomain: "test",
+	},
+}
+
+// Used to mock the API client for the purpose of unit tests.
+type mockAPIClient struct{}
```

Comment:
@jimeh I'd be curious on your thoughts on mocking like this for the CLI, rather than using an interface style mock?

Reply:
I think I would much rather prefer interface mocks via gomock/mockgen or something similar.

While this kind of straight up fake implementation of the `core.RequestMaker` interface works, I think it will become quite tedious and cumbersome to maintain as the CLI tool grows and makes more requests to Katapult's API.

Hence we should probably have an interface that covers the bits of `core.DataCentersClient` that we use, and generate a mock for it via mockgen. Then have each test-case setup the mock with expected calls and return values as needed. I have an example of this in a private repo, so please shout if that might be helpful to see.

Reply:
Yes, I think I agree having an interface probably makes more sense. I'd like to avoid using mockgen and probably just hand code the mock though, it doesn't take long and is a bit more understandable.

Reply:
I would say it depends on how varied the behavior of the mock needs to be, which depends on both how the interface is used, and what is being tested. At a certain point of complexity, the custom mock would need to be flexible enough that it's basically just providing the same feature set as gomock, but with a different/custom API to use it. At that point I'm in favor of just using gomock/mockgen :)

Personally though I'm somewhat undecided, but probably more in favor of just using gomock for consistency, unless it's something very simple which I know will only ever be used in one or two ways. That said, I haven't used gomock all that much yet, or at least not enough to have a bunch of paper cuts from it making me cautious of using it... lol

My primary goal with these kind of things are to try and keep the mock/fake data/logic—however it's defined—tightly coupled to the test case code itself. This avoids long if/else chains elsewhere that cover all manner of test cases across various tests, and it makes it easier to look at a single test case and understand what it does. That said, if lots of test cases setup mocks the same way, I'm happy for helper functions to be used to reduce code duplication, as long as the helper functions have obvious names so it's still reasonably clear what's going on.